File: src/UnwindMC/Generation/Text/CppEmitter.fs
module rec CppEmitter

open System
open System.Collections.Generic
open Ast
open TextWorkflow
open Type
open AstBuilder

let private findRootVar (stmts: IReadOnlyList<Statement>): Var option =
    let rec findRoot =
        function
        | Assignment _ -> None
        | Break -> None
        | DoWhile (loop, _) -> loop |> Seq.tryPick findRoot
        | For (_, modifier, body) -> modifier |> Seq.append body |> Seq.tryPick findRoot
        | FunctionCall _ -> None
        | IfThenElse (_, trueBranch, falseBranch) ->
            let root = trueBranch |> Seq.tryPick findRoot
            if root.IsSome then root else falseBranch |> Seq.tryPick findRoot
        | Return var -> var
        | While (_, loop) -> loop |> Seq.tryPick findRoot
    stmts
    |> Seq.rev
    |> Seq.tryPick findRoot

type private T = {
    func: Function
    parameterNames: HashSet<string>
    types: Dictionary<string, DataType>
}

let emit (func: Function): string =
    let t = {
        func = func
        parameterNames = new HashSet<string>()
        types = new Dictionary<string, DataType>()
    }
    for param in func.parameters do
        t.parameterNames.Add(param.name) |> ignore
        t.types.Add(param.name, param.type_)
    for var in func.locals |> Seq.append func.variables do
        t.types.Add(var.name, var.type_)
    text {
        yield! emitSignature t (findRootVar t.func.body)
        yield! emitBody t
    } |> buildText { indentSize = 2 }

let private emitSignature (t: T) (ret: Var option): TextWorkflow.T =
    text {
        yield! emitReturnType t ret
        yield t.func.name
        yield "("
        for i in [0 .. t.func.parameters.Count - 1] do
            if i <> 0 then
                yield ", "
            yield! emitDeclaration t.func.parameters.[i].name t.func.parameters.[i].type_
        yield ")"
        yield NewLine
    }

let private emitReturnType (t: T) (var: Var option): TextWorkflow.T =
    text {
        match var with
        | None ->
            yield "void "
        | Some (Var name) ->
            let type_ = t.types.[name]
            match type_ with
            | Function ->
                FIXME "emitting functions not supported yet"
            | _ -> ()
            yield "int "
            yield! emitPointerStars type_
    }

let emitPointerStars (type_: DataType): TextWorkflow.T =
    text {
        match type_ with
        | Pointer t ->
            yield "*"
            yield! emitPointerStars t
        | _ -> ()
    }

let private emitBody (t: T): TextWorkflow.T =
    text {
        yield "{"
        yield NewLine
        yield IncreaseIndent

        yield! emitDeclarations t.func.locals t.func.variables
        for node in t.func.body do
            yield! emitStatement t node

        yield DecreaseIndent
        yield "}"
        yield NewLine
    }

let private emitDeclarations (locals: IReadOnlyList<Variable>) (variables: IReadOnlyList<Variable>): TextWorkflow.T =
    text {
        for var in Seq.append locals variables do
            yield! emitDeclaration var.name var.type_
            yield ";"
            yield NewLine
        yield NewLine
    }

let private emitDeclaration (name: string) (type_: DataType): TextWorkflow.T =
    text {
        if isFunction type_ then
            yield "void"
            yield " "
            yield "("
            yield "*"
            yield! emitPointerStars type_
            yield name
            yield ")"
            yield "()"
        else
            yield "int "
            yield! emitPointerStars type_
            yield name
    }

let private emitStatement (t: T) (statement: Statement): TextWorkflow.T =
    text {
        yield!
            match statement with
            | Assignment (var, expr) -> emitAssignment t var expr true
            | Break -> emitBreak ()
            | DoWhile (loop, cond) -> emitDoWhile t loop cond
            | For (cond, modifier, body) -> emitFor t cond modifier body
            | FunctionCall expr -> emitFunctionCall t expr
            | IfThenElse (cond, trueBranch, falseBranch) -> emitIfThenElse t cond trueBranch falseBranch
            | Return var -> emitReturn var
            | While (cond, loop) -> emitWhile t cond loop
    }

let private emitAssignment (t :T) (var: Var) (expr: Expression) (emitSeparator: bool): TextWorkflow.T =
    text {
        let (Var name) = var
        yield name
        yield " = "
        yield! emitExpression t expr
        if emitSeparator then
            yield ";"
            yield NewLine
    }

let private emitBreak (): TextWorkflow.T =
    text {
        yield "break"
        yield ";"
        yield NewLine
    }

let private emitDoWhile (t: T) (loop: IReadOnlyList<Statement>) (cond: Expression): TextWorkflow.T =
    text {
        yield "do"
        yield NewLine

        yield! emitScope t loop true
        yield " "
        yield "while"
        yield " "
        yield "("
        yield! emitExpression t cond
        yield ")"
        yield ";"
        yield NewLine
    }

let private emitFor (t: T) (cond: Expression) (modifier: IReadOnlyList<Statement>) (body: IReadOnlyList<Statement>): TextWorkflow.T =
    text {
        yield "for"
        yield " "
        yield "("
        yield ";"
        yield " "
        yield! emitExpression t cond
        yield ";"
        yield " "
        for (index, stmt) in modifier |> Seq.indexed do
            match stmt with
            | Assignment (var, expr) -> yield! emitAssignment t var expr false
            | _ -> impossible
            if index < modifier.Count - 1 then
                yield ","
                yield " "
        yield ")"
        yield NewLine

        yield! emitScope t body false
    }

let private emitFunctionCall (t: T) (expr: Expression): TextWorkflow.T =
    text {
        yield! emitExpression t expr
        yield "()"
        yield ";"
        yield NewLine
    }

let private emitIfThenElse (t: T) (cond: Expression) (trueBranch: IReadOnlyList<Statement>) (falseBranch: IReadOnlyList<Statement>): TextWorkflow.T =
    text {
        yield "if"
        yield " "
        yield "("
        yield! emitExpression t cond
        yield ")"
        yield NewLine

        yield! emitScope t trueBranch false
        if falseBranch.Count > 0 then
            yield "else"
            yield NewLine
            yield! emitScope t falseBranch false
    }

let private emitReturn (var: Var option): TextWorkflow.T =
    text {
        yield "return"
        match var with
        | Some (Var name) ->
            yield " "
            yield name
        | _ -> ()
        yield ";"
        yield NewLine
    }

let private emitScope (t: T) (stmts: IReadOnlyList<Statement>) (skipNewline: bool): TextWorkflow.T =
    text {
        yield "{"
        yield NewLine

        yield IncreaseIndent
        for node in stmts do
            yield! emitStatement t node
        yield DecreaseIndent

        yield "}"
        if not skipNewline then
            yield NewLine
    }

let private emitWhile (t: T) (cond: Expression) (loop: IReadOnlyList<Statement>): TextWorkflow.T =
    text {
        yield "while"
        yield " "
        yield "("
        yield! emitExpression t cond
        yield ")"
        yield NewLine

        yield! emitScope t loop false
    }

let private emitExpression (t: T) (expression: Expression): TextWorkflow.T =
    text {
        yield!
            match expression with
            | Binary (op, left, right) -> emitBinary t op left right
            | Dereference expr -> emitDereference t expr
            | Unary (op, operand) -> emitUnary t op operand
            | Value value -> emitValue value
            | VarRef var -> emitVar var
    }

let private emitBinary (t: T) (op: Operator) (left: Expression) (right: Expression): TextWorkflow.T =
    text {
        yield! emitExpression t left
        yield " "
        yield
            match op with
            | Operator.Equal -> "=="
            | Operator.NotEqual -> "!="
            | Operator.Less -> "<"
            | Operator.LessOrEqual -> "<="
            | Operator.Greater -> ">"
            | Operator.GreaterOrEqual -> ">="
            | Operator.And -> "&"
            | Operator.Or -> "|"
            | Operator.Xor -> "^"
            | Operator.ShiftLeft -> "<<"
            | Operator.ShiftRight -> ">>"
            | Operator.Add -> "+"
            | Operator.Subtract -> "-"
            | Operator.Multiply -> "*"
            | Operator.Divide -> "/"
            | Operator.Modulo -> "%"
            | _ -> notSupported
        yield " "
        yield! emitExpression t right
    }

let private emitDereference (t: T) (expr: Expression): TextWorkflow.T =
    text {
        yield "*"
        yield "("
        yield! emitExpression t expr
        yield ")"
    }

let private emitUnary (t: T) (op: Operator) (operand: Expression): TextWorkflow.T =
    text {
        yield
            match op with
            | Operator.Negate -> "-"
            | Operator.Not -> "~"
            | _ -> notSupported
        yield! emitExpression t operand
    }

let private emitValue (value: int): TextWorkflow.T =
    text {
        yield value.ToString()
    }

let private emitVar (var: Var): TextWorkflow.T =
    text {
        let (Var name) = var
        yield name
    }

File: src/cpp_emitter.rs
use std::collections::{HashMap, HashSet};

use ast::*;
use ast_builder::*;
use text_writer::*;
use type_resolver::*;

pub struct CppEmitter {
    parameter_names: HashSet<String>,
    types: HashMap<String, DataType>,
    text_writer: TextWriter,
}

impl CppEmitter {
    pub fn emit(func: &Function) -> String {
        let mut emitter = CppEmitter {
            parameter_names: HashSet::new(),
            types: HashMap::new(),
            text_writer: TextWriter::new(2),
        };
        for param in func.parameters.iter() {
            emitter.parameter_names.insert(param.name.clone());
            emitter
                .types
                .insert(param.name.clone(), param.type_.clone());
        }
        for var in func.locals.iter().chain(func.variables.iter()) {
            emitter.types.insert(var.name.clone(), var.type_.clone());
        }
        emitter.emit_signature(func, CppEmitter::find_return_var(&func.body));
        emitter.emit_body(func);
        emitter.text_writer.build_text()
    }

    fn find_return_var(stmts: &Vec<Statement>) -> &Option<Var> {
        // TODO: rewrite Function to include return var
        match stmts.iter().rev().nth(0).unwrap() {
            Statement::Return(var) => var,
            _ => &None,
        }
    }

    fn emit_signature(&mut self, func: &Function, ret: &Option<Var>) {
        self.emit_return_type(ret);
        self.text_writer.write(&func.name);
        self.text_writer.write("(");
        for i in 0..func.parameters.len() {
            if i != 0 {
                self.text_writer.write(",");
                self.text_writer.write(" ");
            }
            self.emit_declaration(&func.parameters[i].name, &func.parameters[i].type_);
        }
        self.text_writer.write(")");
        self.text_writer.write_newline();
    }

    fn emit_return_type(&mut self, var: &Option<Var>) {
        match var {
            None => self.text_writer.write("void"),
            Some(Var::Var(name)) => {
                let type_ = self.types[name].clone();
                match type_ {
                    DataType::Function => panic!("FIXME: emitting functions not supported yet"),
                    _ => (),
                };
                self.text_writer.write("int");
                self.text_writer.write(" ");
                self.emit_pointer_stars(&type_);
            }
        }
    }

    fn emit_body(&mut self, func: &Function) {
        self.text_writer.write("{");
        self.text_writer.write_newline();
        self.text_writer.increase_indent();

        self.emit_declarations(&func.locals, &func.variables);
        for stmt in &func.body {
            self.emit_statement(stmt);
        }

        self.text_writer.decrease_indent();
        self.text_writer.write("}");
        self.text_writer.write_newline();
    }

    fn emit_declarations(&mut self, locals: &Vec<Variable>, variables: &Vec<Variable>) {
        for var in locals.iter().chain(variables.iter()) {
            self.emit_declaration(&var.name, &var.type_);
            self.text_writer.write(";");
            self.text_writer.write_newline();
        }
        self.text_writer.write_newline();
    }

    fn emit_declaration(&mut self, name: &String, type_: &DataType) {
        if type_.is_function() {
            self.text_writer.write("void");
            self.text_writer.write(" ");
            self.text_writer.write("(");
            self.text_writer.write("*");
            self.emit_pointer_stars(type_);
            self.text_writer.write(name);
            self.text_writer.write(")");
            self.text_writer.write("()");
        } else {
            self.text_writer.write("int");
            self.text_writer.write(" ");
            self.emit_pointer_stars(type_);
            self.text_writer.write(name);
        }
    }

    fn emit_pointer_stars(&mut self, type_: &DataType) {
        match type_ {
            DataType::Pointer(t) => {
                self.text_writer.write("*");
                self.emit_pointer_stars(t);
            }
            _ => (),
        }
    }

    fn emit_statement(&mut self, stmt: &Statement) {
        use ast::Statement::*;
        match stmt {
            Assignment(var, expr) => self.emit_assignment(var, expr, true),
            Break => self.emit_break(),
            Continue => self.emit_continue(),
            DoWhile(body, cond) => self.emit_do_while(body, cond),
            For(cond, modifier, body) => self.emit_for(cond, modifier, body),
            FunctionCall(expr) => self.emit_function_call(expr),
            IfThenElse(cond, true_branch, false_branch) => {
                self.emit_if_then_else(cond, true_branch, false_branch)
            }
            Return(var) => self.emit_return(var),
            While(cond, body) => self.emit_while(cond, body),
        }
    }

    fn emit_scope(&mut self, stmts: &Vec<Statement>, skip_newline: bool) {
        self.text_writer.write("{");
        self.text_writer.write_newline();

        self.text_writer.increase_indent();
        for stmt in stmts.iter() {
            self.emit_statement(stmt);
        }
        self.text_writer.decrease_indent();

        self.text_writer.write("}");
        if !skip_newline {
            self.text_writer.write_newline();
        }
    }

    fn emit_assignment(&mut self, var: &Var, expr: &Expression, emit_separator: bool) {
        let Var::Var(name) = var;
        self.text_writer.write(name);
        self.text_writer.write(" ");
        self.text_writer.write("=");
        self.text_writer.write(" ");
        self.emit_expression(expr);
        if emit_separator {
            self.text_writer.write(";");
            self.text_writer.write_newline();
        }
    }

    fn emit_break(&mut self) {
        self.text_writer.write("break");
        self.text_writer.write(";");
        self.text_writer.write_newline();
    }

    fn emit_continue(&mut self) {
        self.text_writer.write("continue");
        self.text_writer.write(";");
        self.text_writer.write_newline();
    }

    fn emit_do_while(&mut self, body: &Vec<Statement>, cond: &Expression) {
        self.text_writer.write("do");
        self.text_writer.write_newline();

        self.emit_scope(body, true);
        self.text_writer.write(" ");
        self.text_writer.write("while");
        self.text_writer.write(" ");
        self.text_writer.write("(");
        self.emit_expression(cond);
        self.text_writer.write(")");
        self.text_writer.write(";");
        self.text_writer.write_newline();
    }

    fn emit_for(&mut self, cond: &Expression, modifier: &Vec<Statement>, body: &Vec<Statement>) {
        self.text_writer.write("for");
        self.text_writer.write(" ");
        self.text_writer.write("(");
        self.text_writer.write(";");
        self.text_writer.write(" ");
        self.emit_expression(cond);
        self.text_writer.write(";");
        self.text_writer.write(" ");
        for (i, stmt) in modifier.iter().enumerate() {
            match stmt {
                Statement::Assignment(var, expr) => self.emit_assignment(var, expr, false),
                _ => panic!("Impossible"),
            }
            if i < modifier.len() - 1 {
                self.text_writer.write(",");
                self.text_writer.write(" ");
            }
        }
        self.text_writer.write(")");
        self.text_writer.write_newline();

        self.emit_scope(body, false);
    }

    fn emit_function_call(&mut self, expr: &Expression) {
        self.emit_expression(expr);
        self.text_writer.write("()");
        self.text_writer.write(";");
        self.text_writer.write_newline();
    }

    fn emit_if_then_else(
        &mut self,
        cond: &Expression,
        true_branch: &Vec<Statement>,
        false_branch: &Vec<Statement>,
    ) {
        self.text_writer.write("if");
        self.text_writer.write(" ");
        self.text_writer.write("(");
        self.emit_expression(cond);
        self.text_writer.write(")");
        self.text_writer.write_newline();

        self.emit_scope(true_branch, false);
        if false_branch.len() > 0 {
            self.text_writer.write("else");
            self.text_writer.write_newline();
            self.emit_scope(false_branch, false);
        }
    }

    fn emit_return(&mut self, var: &Option<Var>) {
        self.text_writer.write("return");
        if let Some(Var::Var(name)) = var {
            self.text_writer.write(" ");
            self.text_writer.write(name);
        }
        self.text_writer.write(";");
        self.text_writer.write_newline();
    }

    fn emit_while(&mut self, condition: &Expression, body: &Vec<Statement>) {
        self.text_writer.write("while");
        self.text_writer.write(" ");
        self.text_writer.write("(");
        self.emit_expression(condition);
        self.text_writer.write(")");
        self.text_writer.write_newline();

        self.emit_scope(body, false);
    }

    fn emit_expression(&mut self, expr: &Expression) {
        use ast::Expression::*;
        match expr {
            Binary(op, left, right) => self.emit_binary(op, left, right),
            Dereference(expr) => self.emit_dereference(expr),
            Unary(op, operand) => self.emit_unary(op, operand),
            Value(value) => self.emit_value(value),
            VarRef(var) => self.emit_var(var),
        }
    }

    fn emit_binary(&mut self, op: &Operator, left: &Expression, right: &Expression) {
        self.emit_expression(left);
        self.text_writer.write(" ");
        use ast::Operator::*;
        self.text_writer.write(match op {
            Equal => "==",
            NotEqual => "!=",
            Less => "<",
            LessOrEqual => "<=",
            Greater => ">",
            GreaterOrEqual => ">=",
            And => "&",
            Or => "|",
            Xor => "^",
            ShiftLeft => "<<",
            ShiftRight => ">>",
            Add => "+",
            Subtract => "-",
            Multiply => "*",
            Divide => "/",
            Modulo => "%",
            _ => panic!("Not supported"),
        });
        self.text_writer.write(" ");
        self.emit_expression(right);
    }

    fn emit_dereference(&mut self, expr: &Expression) {
        self.text_writer.write("*");
        self.text_writer.write("(");
        self.emit_expression(expr);
        self.text_writer.write(")");
    }

    fn emit_unary(&mut self, op: &Operator, operand: &Expression) {
        self.text_writer.write(match op {
            Operator::Negate => "-",
            Operator::Not => "~",
            _ => panic!("Not supported"),
        });
        self.emit_expression(operand);
    }

    fn emit_value(&mut self, value: &i32) {
        self.text_writer.write(&value.to_string());
    }

    fn emit_var(&mut self, var: &Var) {
        let Var::Var(name) = var;
        self.text_writer.write(&name);
    }
}
